File: BAEKJOON/1012.py
import sys
input = sys.stdin.readline
def check(i:int,j:int):
    tempt = [(i,j)]
    dx,dy = [1,-1,0,0],[0,0,-1,1]
    while tempt:
        node = tempt.pop()
        for i in range(4):
            x,y = dx[i] + node[0],dy[i] + node[1]
            if 0<= x < len(matrix) and 0<= y < len(matrix[0]) and not visited[x][y] and matrix[x][y]:
                visited[x][y] = True
                tempt.append((x,y))
for _ in range(int(input())):
    N,M,K = map(int, input().split())
    matrix = [[0 for _ in range(M)] for _ in range(N)]
    for _ in range(K):
        a,b = map(int,input().split())
        matrix[a][b] = 1
    visited = [[False for _ in range(M)]for _ in range(N)]
    cnt = 0
    for i in range(N):
        for j in range(M):
            if matrix[i][j] and not visited[i][j]:
                visited[i][j] = True
                check(i,j)
                cnt += 1
    print(cnt)

File: BAEKJOON/1012.java
import java.io.BufferedReader;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
public class Main {
	static int[] dx = {0,1,0,-1};
	static int[] dy = {1,0,-1,0};
 	public static void main (String[] args) throws IOException {
 		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
 		int N = Integer.parseInt(br.readLine());
 		for(int T = 0; T< N; T ++){
 			String[] input = br.readLine().split(" ");
 			int[][] m = new int[Integer.parseInt(input[0])][Integer.parseInt(input[1])];
 			for(int k = 0; k < Integer.parseInt(input[2]); k++) {
 				String[] in = br.readLine().split(" ");
 				m[Integer.parseInt(in[0])][Integer.parseInt(in[1])] = 1;
 			}
 			int cnt = 0;
 			for(int i = 0;i<m.length; i ++) {
 				for(int j = 0; j < m[0].length; j++) {
 					if(m[i][j] == 1) {
 						cnt ++;
 						m[i][j] = 0;
 						check(i,j,m);
 					}
 				}
 			}
 			System.out.println(cnt);
 			
 			
 		}
	}
 	public static void check(int r, int c, int[][] m) {
 		Queue<int []> q = new LinkedList<>();
 		q.add(new int[] {r,c});
 		while(!q.isEmpty()) {
 			int [] node = q.remove();
 			for(int i = 0; i<4;i ++) {
 				int x = node[0] + dx[i];
 				int y = node[1] + dy[i];
 				if(0<= x && x < m.length && 0<= y && y < m[0].length && m[x][y] == 1) {
 					m[x][y] = 0;
 					q.add(new int[] {x,y});
 				}
 			}
 		}
 	}
 	
}


